core/src/test/java/org/apache/iceberg/TestRowDelta.java

```diff
@@ -1456,6 +1457,58 @@ public void testRewrittenDeleteFiles() {
         statuses(Status.DELETED));
   }
 
+  @TestTemplate
+  public void testRewrittenDeleteFilesReadFromManifest() throws IOException {
+    DataFile dataFile = newDataFile("data_bucket=0");
+    DeleteFile deleteFile = newDeleteFile(dataFile.specId(), "data_bucket=0");
+    RowDelta baseRowDelta = table.newRowDelta().addRows(dataFile).addDeletes(deleteFile);
+    Snapshot baseSnapshot = commit(table, baseRowDelta, branch);
+    assertThat(baseSnapshot.operation()).isEqualTo(DataOperations.OVERWRITE);
+    List<ManifestFile> deleteManifests = baseSnapshot.deleteManifests(table.io());
+    try (ManifestReader<DeleteFile> deleteReader =
+        ManifestFiles.readDeleteManifest(deleteManifests.get(0), table.io(), table.specs())) {
+      deleteFile = deleteReader.iterator().next();
+    }
+
+    assertThat(deleteFile.manifestLocation()).isEqualTo(deleteManifests.get(0).path());
```

Comment:
This test is mostly the same as the one above it, the only difference is we read the delete file from the manifest beforehand so that we hit the new code path.

One thing I went back and forth on was the utility of this test, but I concluded it's better to have it than not. On the one hand, there's no good way to really assert we're exercising the new optimized path since that's all in internal classes (please let me know if I'm missing a way). On the other hand since it is a new path, I felt it's better to have a test which does end up hitting that path and verifies everything is OK.

This combined with the fact that spark rewrite data files action tests exercise the optimized path should give reasonable confidence.